csharp/src/Ice/ConnectionI.cs

```diff
@@ -1447,24 +1450,57 @@ internal void sendHeartbeat()
         {
             if (isActiveOrHolding())
             {
-                OutputStream os = new OutputStream(_instance, Util.currentProtocolEncoding);
-                os.writeBlob(Protocol.magic);
-                Util.currentProtocol.ice_writeMembers(os);
-                Util.currentProtocolEncoding.ice_writeMembers(os);
-                os.writeByte(Protocol.validateConnectionMsg);
-                os.writeByte(0);
-                os.writeInt(Protocol.headerSize); // Message size.
-                try
-                {
-                    sendMessage(new OutgoingMessage(os, false, false));
+                // We check if the connection has become inactive.
+                if (
+                    _inactivityTimer is null &&           // timer not already scheduled
+                    _inactivityTimeout > TimeSpan.Zero && // inactivity timeout is enabled
+                    _state == StateActive &&              // only schedule the timer if the connection is active
+                    _dispatchCount == 0 &&                // no pending dispatch
+                    _asyncRequests.Count == 0 &&          // no pending invocation
+                    _readHeader &&                        // we're not waiting for the remainder of an incoming message
+                    _sendStreams.Count <= 1)              // there is at most one pending outgoing message
+                {
+                    // We may become inactive while the peer is back-pressuring us. In this case, we only schedule the
+                    // inactivity timer if there is no pending outgoing message or the pending outgoing message is a
+                    // heartbeat.
+
+                    // The stream of _sendStreams.First is in _writeStream.
+                    if (_sendStreams.Count == 0 || isHeartbeat(_writeStream))
+                    {
+                        scheduleInactivityTimer();
+                    }
                 }
-                catch (LocalException ex)
+
+                // We send a heartbeat to the peer to generate a "write" on the connection. This write in turns creates
+                // a read on the peer, and resets the peer's idle check timer. When _sendStream.Count > 0, there is
+                // already an outstanding write, so we don't need to send a heartbeat. It's possible _sendStream.First
+                // was sent already but not yet removed from _sendStreams: it means the last write occurred very
+                // recently, which good enough with respect to the idle check.
```

Comment:
which IS good enough

Reply:
Fixed.

Also, I think this is good enough not only because of timing, but because of the logic in IdleTimeoutTransceiverDecorator where we schedule a new heartbeat _after_ a write completes.

When _sendStreams.First corresponds to a just-written frame, there are two possibilities:
 - the IdleTimeoutTransceiverDecorator heartbeat-rescheduling did not happen yet but is about to (=> ok to discard this heartbeat)
 - the IdleTimeoutTransceiverDecorator heartbeat-rescheduling just happened but could not cancel this heartbeat (=> ok to discard it)